README.adoc

```diff
@@ -193,6 +193,48 @@ When a workflow run associated to a pull request is completed, a report is gener
 
 > image::documentation/screenshots/workflow-run-report.png[]
 
+=== Approve workflow runs
+
+This rule applies more fine-grained protections to workflow runs
+than is provided by the basic GitHub settings. If a repository
+is https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository[set up to only allow workflow runs from committers],
+the bot can automatically approve some workflows which meet a set of rules.
+
+Syntax of the `.github/quarkus-github-bot.yml` file is as follows:
+
+[source, yaml]
+----
+features: [ APPROVE_WORKFLOWS ]
+workflows:
+      rules:
+        - allow:
+            directories:
+             - ./src
+             - ./documentation
+             files:
+             - README.md
+            users:
+              minContributions: 5
+        unless:
+           directories:
+             - ./github
```

Comment:
```suggestion
             - .github
```

Reply:
Oh good spot, thank you. The matching on the directory name does need the beginning dot-slash (which is annoying), so I think it should actually be `./.github`. That looks a bit like bad ascii art collided with a product name, but I *think* it's correct. Do you agree?

Reply:
Hmmm. I don't think so. But I think it's the same issue that the one below about `endsWith`/`startsWith` so let's talk there.

Reply:
I'll make a PR and do some logging to confirm the filenames we get from the github api are fully qualified. It's not in the test json in the repo, annoyingly, but I'm sure I saw it go by in my test logging (because I was a bit surprised it was the way it was).

Reply:
Yes they are fully qualified. And I understand now what you’re trying to do. 
I’ll get the PR and have a look at the code tomorrow. 

I missed why you differentiated files and directories. 
My intent with these filters was: either you are pushing a path from the root or you are using a glob.
With files you’re doing something else. But I’m not sure it’s such a good idea because I’m not sure you will be able to be consistent if you want to say exclude all Java files. You will then have to use a full glob with the directory present (well **/*.java) which is not obvious.

Reply:
I’ll have a look tomorrow and think a bit about it. Don’t change anything yet.

Reply:
I agree for *.java you need a glob anyway and then the benefit of having a bit less globbing is marginal. Probably the best use case is for pom.xml where the filename is so standardised. 

Which use case is more likely probably depends on what kind of attacks are more likely. Secret and credential extfiltration would be super-easy with any java file change. Something like bitcoin mining would probably pull in some new dependencies and change the pom.xml (or add something into the workflow, but that's covered by `directories`). https://www.trendmicro.com/vinfo/us/security/news/cybercrime-and-digital-threats/github-action-runners-analyzing-the-environment-and-security-in-action mentions the threats being 
- printing secrets (github filters logs to guard against it)
- cryptomining (github sometimes detects)
... and then it goes very deep and a bit over my head. :)

Reply:
OK so I thought about it a bit more and I think having very different behaviors for files and directories is confusing.

I think I would rather have only one `files` entries with the current directories behavior and rely on glob patterns for files. That would be more consistent and less confusing.
At some point, I would also rename `directories` to `files` for the other rule but it needs careful deployment so I'll do that later specifically.
Also probably a good idea to cache the compiled patterns in a map if we end up using them a lot.
I can do it if you want as it's something I should probably have done before and is a bit orthogonal to your work.

Reply:
Ok, that makes lots of sense! I think renaming to 'files' is the un-locker that makes the behaviour more clear. 

I guess the one thing that wouldn't be entirely obvious in the current/proposed behaviour is 
```
files:
  - ./somedir
```
would match `./somedir/myfile` even though there's no wildcard in it. I can just about make that make sense (to me), though, since a directory *is* a file so `somedir` is a file, and if we include it, of course we include its descendants. 

It makes less sense that
``` 
files:
  - ./somed
```
would match `./somedir/myfile` even though `somed` is not `somedir`. That's orthogonal to this change, though, since it's the existing behaviour. If we were really bothered we could enforce globbing and remove `startsWith`, but I'm not sure it's worth the effort. :) 

I'll make the model change to my rules and see how I go with caching the patterns, since I'm on a roll with caching.

Reply:
https://github.com/quarkusio/quarkus-github-bot/pull/261/commits/edda8676320aaf2f6af399793192f860f03fc834 has the update (reversion of 'files' except that 'directories is now 'files'), and also the caching. 
If nothing else, it's a *lot* less code. :)